core/packages/Condensation.cc

```diff
@@ -0,0 +1,85 @@
+#include "core/packages/Condensation.h"
+#include "core/GlobalState.h"
+
+using namespace std;
+
+namespace sorbet::core::packages {
+
+Condensation::Node &Condensation::pushNode(ImportType type) {
+    auto id = this->nodes_.size();
+    auto isTest = type != ImportType::Normal;
+    auto &node = this->nodes_.emplace_back(id, isTest);
+    return node;
+}
+
+const Condensation::Traversal Condensation::computeTraversal(const core::GlobalState &gs) const {
+    Traversal result;
+
+    // The nodes we're currently exploring, and will explore on the next iteration.
+    vector<uint32_t> frontier;
+    vector<uint32_t> next;
+
+    vector<int> neededImports(this->nodes_.size(), 0);
+    vector<vector<uint32_t>> backEdges(this->nodes_.size());
+
+    // Seed the needed imports for the traversal from the roots.
+    for (auto &node : this->nodes_) {
+        neededImports[node.id] = node.imports.size();
+
+        if (node.imports.empty()) {
+            frontier.emplace_back(node.id);
+        }
+
+        for (auto imp : node.imports) {
+            backEdges[imp].push_back(node.id);
+        }
+    }
+
+    // The reservations for `result.packages` and `result.sccs` are important: we are constructing spans into both
+    // vectors as we traverse the graph, and a reallocation will invalidate those spans.
+    result.packages.reserve(2 * gs.packageDB().packages().size());
+    result.sccs.reserve(this->nodes_.size());
```

Comment:
This is a little fragile, and I'm not sure what a better approach looks like. We could switch to storing start/length pairs in `SCCInfo` and the `parallel` vector, but it seemed nice to have the spans ready to go.